solr/core/src/java/org/apache/solr/store/blob/metadata/CorePushPull.java

```diff
@@ -268,7 +264,7 @@ public void pullUpdateFromBlob(long requestQueuedTimeMs, boolean waitForSearcher
                   }
 
```

Comment:
Can't seem to comment on the unchanged lines, therefore, commenting here:
https://github.com/apache/lucene-solr/pull/1055/files?utf8=%E2%9C%93&diff=unified&w=1#diff-d86585f4b79b136f1575702bd7c61e38R226 is not correct, it talks about incorrect understanding of directory locking.

Pull is now an exclusive operation(i.e.no other pull or active indexing should be happening during this time). If local directory contents have changed we fail the pull. Therefore:

1. After closing index writer we should obtain an exclusive lock on directory, similar to https://github.com/apache/lucene-solr/blob/1133bf98a5fd075173efecfb75a51493fceb62b3/solr/core/src/java/org/apache/solr/update/SolrIndexSplitter.java#L161

2. We should close index write and acquire write lock in the beginning of method so that the chances of directory changes(most likely background merges?) are even less.

3. The only purpose of temp directory, I see, is to prevent a partial download of index(transient network error) from corrupting the index directory. Otherwise we could have just downloaded directly into index directory. Comments on temp directory usage need to be updated to reflect that.

Reply:
In general, Solr does not rely on the low level lucene IndexWriter lock.  It's really just an additional safety mechanism, and a failure to aquire the lock really means that something else went badly wrong (a bug, or an external process.) The one true IndexWriter (and changes to it) are generally controlled by the SolrCoreState class.

The comments around pull being an exclusive operation sound correct.

Reply:
Pull is indeed exclusive, but let's not rely on this (i.e. be defensive). We do check the directory hasn't changed during the pull before adding back the files and reopening the IW, so I think we're ok.

Not sure about your reference @mbwaheed to SolrIndexSplitter. The lock is acquired there on a directory that I'm not sure is the index directory (it is searcher.getRawReader().directory() on the passed instance of SolrIndexSearcher). We manipulate solrCore.getIndexDir().

The directory and IndexWriter manipulation in IndexFetcher.fetchLatestIndex() is similar to the one we do here (no surprise, it was used as "inspiration").

Reply:
I read this and your updated comments in the code. Sounds good.